### DISCUSSION TITLE
Doubts about comms with running bevy simulation

### DISCUSSION THREAD

**Discussion by karnotxo** [bevyengine/bevy#3964] [Q&A]
I'm still starting with rust and just discovered about ECS's
I'm trying to build kinda physics simulation and decided to use bevy_ecs... not a game nor need  to display graphics, so I think it's ok... I'm encapsulating sim stuff in a lib and need this to generate soem udp messages as output. Also, I'm planning to have two options for launchinh. First from command line,  with a few parameters and just launching simulation... or also from a minimal ui, that should display some statistics from the simulation in real time. I think I have idea about building any of these modules alone but not sure about how to build communication with muy  running simulation, so:
1) any thoughts about how to implement a system that sends this UDP messaging (also, data sent by messaging includes a lot more info  than what is strictly needed for running our simulation, so this info musdt be taken from somewhere else... perhaps a database?)
2) how should I do for querying from this minimal UI to my running simulation so I can show some statistics in my widgets? (I guess simulation should run in a different thread from UI)
Thanks a lot in advance!

**Comment by lukors** [comment]
Answers to questions:

1. When you have the data you want to send you generate messages and send it. I'm probably misunderstanding the question. I don't understand what you've written in the parenthesis, so that might be where my problem lies.
2. The way I've solved this in my application is that the library runs in a thread, and it writes information to a place accessible from the outside, which I can then read from Bevy without blocking. I'm using `RwLock`s for this.

**Reply by karnotxo** [reply]
Thanks Lukors. I edited my question so what I wrote into parentheses has soem sense now... Sorry for that :/
I'm afraid writing to a file for data exchange would be not fast enough... Perhaps some king of message qeue or synchronized memory based struct could be an option?

**Reply by lukors** [reply]
Thank you for clarifying question 1! :)

I think the UDP stuff is probably too advanced for me as I still don't understand it. So I don't think I can help there.

---
I don't mean writing it to a file, I mean writing it to a place in memory that can be reached from outside the thread. `RwLock` lets you do stuff like have a struct that can be reached from two different threads.

So you could have somehting like `Arc<RwLock<MyStruct>>`, then you just clone the `Arc` and give that to the thread when you create it, and that enables you to lock the `RwLock` to read or write to it from both threads whenever you want. :)
